File: killstreek/src/main/java/dev/kkorolyov/killstreek/component/Health.java
package dev.kkorolyov.killstreek.component;

import dev.kkorolyov.killstreek.utility.BoundedValue;
import dev.kkorolyov.pancake.platform.entity.Component;

/**
 * Maintains an entity's state of existence.
 */
public class Health implements Component {
	private BoundedValue<Integer> value;

	/**
	 * Constructs a new health with initial health set to {@code max}.
	 * @param max maximum value for current health, constrained {@code > 0}
	 */
	public Health(int max) {
		this(max, max);
	}
	/**
	 * Constructs a new health.
	 * @param max maximum value for current health, constrained {@code > 0}
	 * @param current initial health, constrained {@code <= max}
	 */
	public Health(int max, int current) {
		value = new BoundedValue<>(null, max, current);
	}

	/**
	 * Applies a change to current health.
	 * @param amount change in current health
	 */
	public void change(int amount) {
		value.set(value.get() + amount);
	}

	/** @return {@code true} if current health {@code <= 0} */
	public boolean isDead() {
		return value.get() <= 0;
	}
	/** @return {@code true} if current health {@code < 0} */
	public boolean isSuperDead() {
		return value.get() < 0;
	}

	/** @return health value */
	public BoundedValue<Integer> getValue() {
		return value;
	}
}

File: killstreek/src/main/kotlin/dev/kkorolyov/killstreek/component/Health.kt
package dev.kkorolyov.killstreek.component

import dev.kkorolyov.pancake.platform.entity.Component
import dev.kkorolyov.pancake.platform.math.BoundedValue

/**
 * Maintains an entity's state of existence.
 * @param max maximum value for health, constrained {@code > 0}
 * @param current initial health, constrained {@code <= max}
 */
class Health(max: Int, current: Int = max) : Component {
	/** health value */
	val value: BoundedValue<Int> = BoundedValue<Int>(null, max, current)

	/** percentage of current health with respect to max health */
	val percentage: Double get() = value.get().toDouble() / value.maximum.toDouble()

	/** whether current health {@code <= 0} */
	val isDead: Boolean get() = value.get() <= 0
	/** whether current health {@code < 0} */
	val isSuperDead: Boolean get() = value.get() < 0

	/**
	 * Applies a change to current health.
	 * @param amount change in current health
	 */
	fun change(amount: Int) {
		value.set(value.get() + amount)
	}
}
